crates/forge_app/src/compact.rs

```diff
@@ -145,3 +146,249 @@ impl<S: AgentService> Compactor<S> {
+        // 14.4
+        let actual = seq("sutrtru", 0.4);
+        assert_eq!(actual, "sutrtru");
+
+        // Test with very small percentage
+        // Pattern: s-u-a-a-a-a, Total: u(3) + a(5) + a(5) + a(5) + a(5) = 23 tokens
+        // 10% of 23 = 2.3 tokens, Groups: [u](3) > 2.3, so no groups fit
+        let actual = seq("suaaaa", 0.1);
+        assert_eq!(actual, "suaaaa");
+
+        // Test with pattern starting with user message (no system)
+        // Pattern: u-a-t-r-u, Total: u(3) + a(5) + tr(15) + u(3) = 26 tokens
+        // 40% of 26 = 10.4 tokens, Groups: [u](3) + [a](5) = 8 < 10.4, but adding
+        // [tr](15) = 23 > 10.4
+        let actual = seq("uatru", 0.4);
+        assert_eq!(actual, "[ua]tru");
+
+        // Test with all assistant messages
+        // Pattern: s-a-a-a-a, Total: a(5) + a(5) + a(5) + a(5) = 20 tokens
+        // 60% of 20 = 12 tokens, Groups: [a](5) + [a](5) = 10 < 12, but adding [a](5) =
+        // 15 > 12
+        let actual = seq("saaaa", 0.6);
+        assert_eq!(actual, "s[aa]aa");
+
+        // Test complex pattern with tool calls - each tr is a group
+        // Pattern: s-u-t-r-t-r-a-a, Total: u(3) + tr(15) + tr(15) + a(5) + a(5) = 43
+        // tokens 50% of 43 = 21.5 tokens, Groups: [u](3) + [tr](15) = 18 <
+        // 21.5, but adding [tr](15) = 33 > 21.5
+        let actual = seq("sutrtraa", 0.5);
+        assert_eq!(actual, "s[utr]traa");
+
+        // Test pattern with single non-system message
+        // Pattern: s-u, Total: u(3) = 3 tokens
+        // 50% of 3 = 1.5 tokens, Groups: [u](3) > 1.5, so no groups fit
+        let actual = seq("su", 0.5);
+        assert_eq!(actual, "su");
+
+        // Test pattern with alternating messages
+        // Pattern: s-u-a-u-a-u-a, Total: u(3) + a(5) + u(3) + a(5) + u(3) + a(5) = 24
+        // tokens 30% of 24 = 7.2 tokens, Groups: [u](3) < 7.2, but adding
+        // [a](5) = 8 > 7.2
+        let actual = seq("suauaua", 0.3);
+        assert_eq!(actual, "suauaua");
+
+        // Test with tool call followed by multiple results (if that's possible)
+        // Pattern: s-u-t-r-u, Total: u(3) + tr(15) + u(3) = 21 tokens
+        // 50% of 21 = 10.5 tokens, Groups: [u](3) < 10.5, but adding [tr](15) = 18 >
+        // 10.5
+        let actual = seq("sutru", 0.5);
+        assert_eq!(actual, "sutru");
+
+        // Test where tool call group fits within percentage
+        // Pattern: s-u-t-r, Total: u(3) + tr(15) = 18 tokens
+        // 90% of 18 = 16.2 tokens, Groups: [u](3) < 16.2, but adding [tr](15) = 18 >
+        // 16.2
+        let actual = seq("sutr", 0.9);
+        assert_eq!(actual, "sutr");
+
+        // Test where tool call group actually fits within percentage
+        // Pattern: s-u-t-r, Total: u(3) + tr(15) = 18 tokens
+        // 100% of 18 = 18 tokens, Groups: [u](3) + [tr](15) = 18 = 18, so both fit
+        // exactly
+        let actual = seq("sutr", 1.0);
+        assert_eq!(actual, "s[utr]");
+    }
+}
```

Comment:
We could consider making each item in the message seq to be 10 tokens long. That way the math will become simpler and expected values will become easier to guess.